hpx/traits/segmented_iterator_traits.hpp

```diff
@@ -50,6 +50,172 @@ namespace hpx { namespace traits
+
+                template <typename SegIter>
+                typename result<segmented_iterator_begin(SegIter)>::type
+                operator()(SegIter iter) const
+                {
+                    return segmented_iterator_traits<Iterator>::begin(iter);
+                };
+            };
+        };
+
+        struct segmented_iterator_end
+        {
+            template <typename Iterator>
+            struct apply
+            {
+                typedef typename segmented_iterator_traits<
+                        Iterator
+                    >::local_iterator type;
+
+                template <typename T>
+                struct result;
+
+                template <typename This, typename SegIter>
+                struct result<This(SegIter)>
+                {
+                    typedef typename apply::type type;
+                };
+
+                template <typename SegIter>
+                typename result<segmented_iterator_end(SegIter)>::type
+                operator()(SegIter iter) const
+                {
+                    return segmented_iterator_traits<Iterator>::end(iter);
+                };
+            };
+        };
+
+        struct segmented_iterator_local_begin
+        {
+            template <typename Iterator>
+            struct apply
+            {
+                typedef typename segmented_iterator_traits<
+                        Iterator
+                    >::local_raw_iterator type;
+
+                template <typename T>
+                struct result;
+
+                template <typename This, typename LocalSegIter>
+                struct result<This(LocalSegIter)>
+                {
+                    typedef typename apply::type type;
+                };
+
+                template <typename LocalSegIter>
+                typename result<segmented_iterator_local_begin(LocalSegIter)>::type
+                operator()(LocalSegIter iter) const
+                {
+                    return segmented_iterator_traits<Iterator>::begin(iter);
+                };
+            };
+        };
+
+        struct segmented_iterator_local_end
+        {
+            template <typename Iterator>
+            struct apply
+            {
+                typedef typename segmented_iterator_traits<
+                        Iterator
+                    >::local_raw_iterator type;
+
+                template <typename T>
+                struct result;
+
+                template <typename This, typename LocalSegIter>
+                struct result<This(LocalSegIter)>
+                {
+                    typedef typename apply::type type;
+                };
+
+                template <typename LocalSegIter>
+                typename result<segmented_iterator_local_end(LocalSegIter)>::type
+                operator()(LocalSegIter iter) const
+                {
+                    return segmented_iterator_traits<Iterator>::end(iter);
+                };
+            };
+        };
+    }
```

Comment:
Those function objects have been moved from the `zip_iterator` file to here. Initially I was planning to enable the stencil iterator to be usable in distributed contexts (for which I needed those function objects). Their purpose is to lift the transform all iterators wrapped by a zip_iterator in order to create a new zip_iterator wrapping the transformed iterators (see for instance: https://github.com/STEllAR-GROUP/hpx/blob/transform_iterator/hpx/util/zip_iterator.hpp#L579).